In [399]:
import pandas as pd
import os
os.chdir("C:/Users/likhi/OneDrive/Desktop/DataSets")

# Details about my Project
I have made a survey and collected data about daily internet usage by mobile and web applications
This survey is made to analyse the data that is created everyday.I asked how much time does they spend on some apps,which are most used and most popular among the indian young generation.

By using a data usage app, i computed data consumed by each app in a minute.I tested out each app for 15 minutes and gave the average data usage in a minute

#### My analysis is as follows:

Facebook       -     24 MB/minute(Inclusive of all features)<br>
>1 like/comment     -   25 KB<br>
1 share     - 1 250 KB<br>
HomeFeed    -      8 MB/minute<br>
   Videos      -     40 MB/minute<br>
   Gaming      -     40 MB/minute<br>
   
Instagram      -     20 MB/minute(Inclusive of all features)<br>
   >HomeFeed    -     20 MB/minute<br>
   Reels       -     20 MB/minute<br>
   others      -     25 MB/minute<br>
   1 like/comment      -     25 KB<br>
   1 share       -      1 MB<br>
   
Whatsapp       -   0.20 MB/minute<br>
>1 msg    -    1 KB<br>
1 minute call   -   0.25 MB<br>
1 hour call     -   24 MB
5 minutes in whatsapp   -   1 MB or less

Youtube        -    8 MB/minute<br>
>1 hour in youtube  - 450 MB<br>

Streaming Services    -    8 MB/minute<br>
>Netflix ------Good quality- 0.70 GB/Hour<br>
 --------------Best Quality- 3 GB/Hour<br>
 Amazon Prime--Good quality-0.18 GB<br>
 --------------Best Quality - 0.52 GB<br>

Twitter       -     4 MB/minute<br>
Spotify/Gaana/oth.    -    2 MB/minute<br>
Snapchat      -     10 MB/minute<br>
PUBG/other games   -   1 MB/minute<br>
Gmail         -      0.20 MB/minute
UPI Apps      -      5 MB/transaction
1 webpage     -      170 KB to 333 KB

   

In [400]:
headers=['time','email','gender','location','facebook','fb_feature','instagram','insta_feature','whatsapp','google','youtube',
         'snapchat','twitter','music','gmail','ott','ott_week','pubg','pubg_tier','web','apps','os','others']
df=pd.read_csv("Daily-Data-Usage-Statistics-Survey.csv")
df.columns=headers
df=df.drop(['time','email','location'],axis=1)
df.describe(include='all')

gender facebook fb_feature   instagram insta_feature    whatsapp  \
count      71       66         43   69.000000            62   70.000000   
unique      3       23          6         NaN             4         NaN   
top      Male        0  Home Feed         NaN     Home Feed         NaN   
freq       49       28         22         NaN            51         NaN   
mean      NaN      NaN        NaN   74.652174           NaN  105.414286   
std       NaN      NaN        NaN  113.582935           NaN  142.471710   
min       NaN      NaN        NaN    0.000000           NaN    2.000000   
25%       NaN      NaN        NaN   20.000000           NaN   32.250000   
50%       NaN      NaN        NaN   54.000000           NaN   60.000000   
75%       NaN      NaN        NaN  120.000000           NaN  120.000000   
max       NaN      NaN        NaN  880.000000           NaN  880.000000   

            google      youtube    snapchat     twitter        music  \
count     69.00000    70.000000   66.000000   64.000000    68.000000   
unique         NaN          NaN         NaN         NaN          NaN   
top            NaN          NaN         NaN         NaN          NaN   
freq           NaN          NaN         NaN         NaN          NaN   
mean      86.42029   110.485714    7.181818   13.515625    56.691176   
std      176.27572   197.163490   19.717941   39.324451   157.372384   
min        0.00000     0.000000    0.000000    0.000000     0.000000   
25%       15.00000    21.000000    0.000000    0.000000     0.000000   
50%       30.00000    60.000000    0.000000    1.000000    15.000000   
75%       70.00000   120.000000    0.000000   10.000000    30.000000   
max     1200.00000  1500.000000  120.000000  250.000000  1200.000000   

             gmail        ott    ott_week        pubg pubg_tier          web  \
count    69.000000   67.00000   64.000000   65.000000        42    62.000000   
unique         NaN        NaN         NaN         NaN        26          NaN   
top            NaN        NaN         NaN         NaN         0          NaN   
freq           NaN        NaN         NaN         NaN         8          NaN   
mean     14.420290   83.19403   52.742188   45.338462       NaN   183.435484   
std      29.666978   89.89240  104.520067   92.044024       NaN   677.893232   
min       0.000000    0.00000    0.000000    0.000000       NaN     0.000000   
25%       3.000000    0.00000    1.750000    0.000000       NaN     0.000000   
50%      10.000000   60.00000    6.000000    0.000000       NaN    25.000000   
75%      15.000000  150.00000   57.000000   20.000000       NaN   115.000000   
max     240.000000  300.00000  500.000000  400.000000       NaN  5200.000000   

              apps       os others  
count    68.000000       70     19  
unique         NaN        2     16  
top            NaN  Android      0  
freq           NaN       63      4  
mean      6.794118      NaN    NaN  
std      20.863135      NaN    NaN  
min       0.000000      NaN    NaN  
25%       1.000000      NaN    NaN  
50%       2.000000      NaN    NaN  
75%       5.000000      NaN    NaN  
max     160.000000      NaN    NaN

# Studying the survey data
From the about description of the dataframes, we can assume that the data is not precise to what is asked.So we need to format the data and replace improper datatypes.

we need to study each of the acquired object, so we follow this below process.
1.Replace and Format the data and its datatype
2.Acquire the Mean time of the data and standard deviation of the data using normalization.
3.Compare the usage of the app with gender.
4.compute required addditional computations

***
## Obseravtion on facebook


In [401]:
df['facebook'].describe()

count     66
unique    23
top        0
freq      28
Name: facebook, dtype: object

## Data format function:
Here is the data format function which we use to determine missing items or incorrect entries and replace them.
Also the time in minutes are considered at max 300 for any app.
Any data item exceeding 300 will resukt in the data to be 300 only,if any data item exceeds 600 which is almost 10 hours , it will be omitted and considered as a fault data entry.

In [402]:
def dataformat(obj):
    obj=obj.fillna(0)
    data_object=[]
    for each in obj:
        each=str(each)
        if(each.isnumeric()):
            each=int(each)
        else:
            k=''
            for letter in each:
                if(letter.isnumeric()):
                    k=k+letter
            if(not k):
                each=0
            else:
                each=int(k)
        if(each>300 and each<=600):
            each=300
        if(each>600):
            each=0
        data_object.append(each)
    obj=data_object
    return obj

In [403]:
obj=dataformat(df['facebook'])
df['facebook']=obj
df['facebook'].describe()

count     71.000000
mean      19.704225
std       46.184226
min        0.000000
25%        0.000000
50%        0.000000
75%       15.000000
max      300.000000
Name: facebook, dtype: float64

#### observing Instagram

In [404]:
obj=dataformat(df['instagram'])
df['instagram']=obj
df['instagram'].describe()

count     71.000000
mean     129.014085
std      138.884895
min        0.000000
25%        0.000000
50%       30.000000
75%      300.000000
max      300.000000
Name: instagram, dtype: float64

#### observing whatsapp

In [405]:
obj=dataformat(df['whatsapp'])
df['whatsapp']=obj
df['whatsapp'].describe()

count     71.000000
mean     135.774648
std      142.304121
min        0.000000
25%        0.000000
50%       50.000000
75%      300.000000
max      300.000000
Name: whatsapp, dtype: float64

#### observing Google/Chrome

In [406]:
obj=dataformat(df['google'])
df['google']=obj
df['google'].describe()

count     71.000000
mean     151.549296
std      132.639661
min        0.000000
25%        0.000000
50%      150.000000
75%      300.000000
max      300.000000
Name: google, dtype: float64

#### observing Youtube

In [407]:
obj=dataformat(df['youtube'])
df['youtube']=obj
df['youtube'].describe()

count     71.000000
mean     116.478873
std      132.299429
min        0.000000
25%        0.000000
50%        0.000000
75%      300.000000
max      300.000000
Name: youtube, dtype: float64

#### Observing Twitter

In [408]:
obj=dataformat(df['twitter'])
df['twitter']=obj
df['twitter'].describe()

count     71.000000
mean      50.366197
std       81.611141
min        0.000000
25%        0.000000
50%       10.000000
75%       50.000000
max      300.000000
Name: twitter, dtype: float64

#### Observing Snapchat

In [409]:
obj=dataformat(df['snapchat'])
df['snapchat']=obj
df['snapchat'].describe()

count     71.000000
mean      37.183099
std       86.142458
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max      300.000000
Name: snapchat, dtype: float64

#### Observing Music Apps

In [410]:
obj=dataformat(df['music'])
df['music']=obj
df['music'].describe()

count     71.000000
mean     105.070423
std      126.895607
min        0.000000
25%        0.000000
50%       20.000000
75%      250.000000
max      300.000000
Name: music, dtype: float64

#### Observing Gmail

In [411]:
obj=dataformat(df['gmail'])
df['gmail']=obj
df['gmail'].describe()

count     71.000000
mean     100.140845
std       91.190428
min        0.000000
25%       20.000000
50%      100.000000
75%      150.000000
max      300.000000
Name: gmail, dtype: float64

#### Observing Streaming Services

In [412]:
obj=dataformat(df['ott'])
df['ott']=obj
df['ott'].describe()

count     71.000000
mean      59.154930
std      116.591429
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max      300.000000
Name: ott, dtype: float64

Since ott_week data is biased by 50% it is vomitted,
the pubg tier information is also not essential

Hence we drop those fields.

In [413]:
df=df.drop(['ott_week','pubg_tier'],axis=1)

###### For web application and apps downloads data, we need to change data format function

In [414]:
obj=[]
for i in df['web']:
    if(i>720):
        i=720
    obj.append(i)
df['web']=obj
df['web'].describe()

count     62.000000
mean     102.467742
std      175.383768
min        0.000000
25%        0.000000
50%       25.000000
75%      115.000000
max      720.000000
Name: web, dtype: float64

In [415]:
obj=[]
for i in df['apps']:
    if(i>50):
        i=50
    obj.append(i)
df['apps']=obj
df['apps'].describe()

count    68.000000
mean      5.029412
std       9.759757
min       0.000000
25%       1.000000
50%       2.000000
75%       5.000000
max      50.000000
Name: apps, dtype: float64

#### A bit strange but wrong computation.

It is time to calculate data generated by the mobile and web applications.<br>
We find mean and multiply.

In [416]:
mean_data=df.mean()
mean_data

facebook      19.704225
instagram    129.014085
whatsapp     135.774648
google       151.549296
youtube      116.478873
snapchat      37.183099
twitter       50.366197
music        105.070423
gmail        100.140845
ott           59.154930
pubg          45.338462
web          102.467742
apps           5.029412
dtype: float64

In [428]:
usage=[8,10,0.20,0.30,10,8,4,2,0.20,10,1,0.30,0]
total_data=0
for each in range(len(mean_data)):
    total_data+=(mean_data[each]*usage[each])
print(total_data)

4081.9097700346


# Really?
<br>
4081 MB, which is 4 GB of data is the average data used by a person in a day.<br>
But something doesnt workout right? Most of the mobile data users use upto 2 GB a day,but getting an average of 4 GB is not correct.This happens because we have calculated the wrong approach.<br>

Lets consider PUBG, less than 50% play PUBG but when we calculate average and compute the result, it is granted that everyone plays PUBG for an average time, which is completely wrong.<br>

Hence our approach should be calculating Data used by each user and computing their average.<br>

Lets Do It

In [429]:
new_df=df.drop(['gender','fb_feature','insta_feature','os','others','apps'],axis=1)
new_df.describe()

facebook   instagram    whatsapp      google     youtube    snapchat  \
count   71.000000   71.000000   71.000000   71.000000   71.000000   71.000000   
mean    19.704225  129.014085  135.774648  151.549296  116.478873   37.183099   
std     46.184226  138.884895  142.304121  132.639661  132.299429   86.142458   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000   30.000000   50.000000  150.000000    0.000000    0.000000   
75%     15.000000  300.000000  300.000000  300.000000  300.000000    0.000000   
max    300.000000  300.000000  300.000000  300.000000  300.000000  300.000000   

          twitter       music       gmail         ott        pubg         web  
count   71.000000   71.000000   71.000000   71.000000   65.000000   62.000000  
mean    50.366197  105.070423  100.140845   59.154930   45.338462  102.467742  
std     81.611141  126.895607   91.190428  116.591429   92.044024  175.383768  
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000  
25%      0.000000    0.000000   20.000000    0.000000    0.000000    0.000000  
50%     10.000000   20.000000  100.000000    0.000000    0.000000   25.000000  
75%     50.000000  250.000000  150.000000    0.000000   20.000000  115.000000  
max    300.000000  300.000000  300.000000  300.000000  400.000000  720.000000

We Calculate the total time(in minutes) and later compute the wholesum of data used

In [430]:
new_df['total_usage_time']=df.sum(axis=1)
new_df

facebook  instagram  whatsapp  google  youtube  snapchat  twitter  music  \
0        300          0        50      50        0         0       10     10   
1          0          0         0     100        0       100       10      0   
2         10        100         0       0        0         0        0      0   
3         10          0         0     300        0       300      100    300   
4          0          0       300     300        0         0      200    200   
..       ...        ...       ...     ...      ...       ...      ...    ...   
66        10          0         0       0        0       300        0      0   
67         0        300       300       0      300         0        0      0   
68        30        300         0     150      300         0       50    150   
69        15          0       300     300      150         0        0    300   
70        60          0         0       0        0         0        0     20   

    gmail  ott   pubg    web  total_usage_time  
0      20    0    1.0    1.0             450.0  
1     100    0    0.0   60.0             371.0  
2     300    0    0.0  500.0             918.0  
3     300    0    0.0   10.0            1328.0  
4     150    0    0.0   60.0            1212.0  
..    ...  ...    ...    ...               ...  
66      0    0  180.0    0.0             491.0  
67      0    0    0.0    0.0             900.0  
68      0    0  180.0    0.0            1162.0  
69      0    0    0.0    0.0            1065.0  
70     10    0  120.0   60.0             271.0  

[71 rows x 13 columns]

Since we included a new column, lets include a new usage constant so that the data is preserved.<br>
We replace the apps usage and replace it ...Also apps data is not bery useful,hence it is vomitted.

In [431]:
usage[-1]=1

In [432]:
index=0
for i in new_df:
    #print(usage[index],i)
    new_df[i]*=usage[index]
    index+=1
new_df

facebook  instagram  whatsapp  google  youtube  snapchat  twitter  music  \
0       2400          0      10.0    15.0        0         0       40     20   
1          0          0       0.0    30.0        0       800       40      0   
2         80       1000       0.0     0.0        0         0        0      0   
3         80          0       0.0    90.0        0      2400      400    600   
4          0          0      60.0    90.0        0         0      800    400   
..       ...        ...       ...     ...      ...       ...      ...    ...   
66        80          0       0.0     0.0        0      2400        0      0   
67         0       3000      60.0     0.0     3000         0        0      0   
68       240       3000       0.0    45.0     3000         0      200    300   
69       120          0      60.0    90.0     1500         0        0    600   
70       480          0       0.0     0.0        0         0        0     40   

    gmail  ott   pubg    web  total_usage_time  
0     4.0    0    1.0    0.3             450.0  
1    20.0    0    0.0   18.0             371.0  
2    60.0    0    0.0  150.0             918.0  
3    60.0    0    0.0    3.0            1328.0  
4    30.0    0    0.0   18.0            1212.0  
..    ...  ...    ...    ...               ...  
66    0.0    0  180.0    0.0             491.0  
67    0.0    0    0.0    0.0             900.0  
68    0.0    0  180.0    0.0            1162.0  
69    0.0    0    0.0    0.0            1065.0  
70    2.0    0  120.0   18.0             271.0  

[71 rows x 13 columns]

We add up the sum of row axis and form a new DataSum column, But we need to exclude the total_usage_time.<br>
Hence The equation will be as follows.

In [433]:
new_df['sum']=new_df.sum(axis=1)-new_df['total_usage_time']
new_df

facebook  instagram  whatsapp  google  youtube  snapchat  twitter  music  \
0       2400          0      10.0    15.0        0         0       40     20   
1          0          0       0.0    30.0        0       800       40      0   
2         80       1000       0.0     0.0        0         0        0      0   
3         80          0       0.0    90.0        0      2400      400    600   
4          0          0      60.0    90.0        0         0      800    400   
..       ...        ...       ...     ...      ...       ...      ...    ...   
66        80          0       0.0     0.0        0      2400        0      0   
67         0       3000      60.0     0.0     3000         0        0      0   
68       240       3000       0.0    45.0     3000         0      200    300   
69       120          0      60.0    90.0     1500         0        0    600   
70       480          0       0.0     0.0        0         0        0     40   

    gmail  ott   pubg    web  total_usage_time     sum  
0     4.0    0    1.0    0.3             450.0  2490.3  
1    20.0    0    0.0   18.0             371.0   908.0  
2    60.0    0    0.0  150.0             918.0  1290.0  
3    60.0    0    0.0    3.0            1328.0  3633.0  
4    30.0    0    0.0   18.0            1212.0  1398.0  
..    ...  ...    ...    ...               ...     ...  
66    0.0    0  180.0    0.0             491.0  2660.0  
67    0.0    0    0.0    0.0             900.0  6060.0  
68    0.0    0  180.0    0.0            1162.0  6965.0  
69    0.0    0    0.0    0.0            1065.0  2370.0  
70    2.0    0  120.0   18.0             271.0   660.0  

[71 rows x 14 columns]

In [438]:
new_df[['sum','total_usage_time']].describe()

sum  total_usage_time
count     71.000000         71.000000
mean    4074.181690       1040.239437
std     2654.926811        454.014269
min      230.000000        106.000000
25%     2195.000000        744.500000
50%     3633.000000       1032.000000
75%     5946.500000       1285.000000
max    11126.500000       2176.000000

Almost There, We got the Data used in a day and time spent on internet.<br>
Hence we need to divide the object properties to obtain the optimal result

In [435]:
new_df['data_per_minute']=new_df['sum']/new_df['total_usage_time']

In [436]:
new_df['data_per_minute'].describe()

count    71.000000
mean      3.821501
std       1.857032
min       0.477004
25%       2.441432
50%       3.776986
75%       5.310270
max       7.677419
Name: data_per_minute, dtype: float64

In [437]:
new_df['data_per_minute'].mean()

3.821500534160072

#### Hence The Data Used by a Person in a minute is 3.8 MB.

But wait, this is just the data usage analysis, What about the data generated in the backend?<br>
What about the data stored and retrieved by the respective company ? Tons of Data is created by each person unknown to themselves, What about it?<br>


....Continuing my research on backend data analytics<br>
....open for suggestions